FLT/Deformations/RepresentationTheory/Subrepresentation.lean

```diff
@@ -0,0 +1,74 @@
+import Mathlib.RepresentationTheory.Basic
+import FLT.Mathlib.LinearAlgebra.Span.Defs
+
+open Pointwise
+
+universe u
+
+variable {A : Type*} [CommRing A]
+
+variable {G : Type*} [Group G]
+
+variable {W : Type*} [AddCommMonoid W] [Module A W]
+
+variable {ρ : Representation A G W}
+
+variable (ρ) in
+structure Subrepresentation where
+  toSubmodule : Submodule A W
+  apply_mem_toSubmodule (g : G) ⦃v : W⦄ : v ∈ toSubmodule → ρ g v ∈ toSubmodule
+
+namespace Subrepresentation
+
+lemma toSubmodule_injective : Function.Injective (toSubmodule : Subrepresentation ρ → Submodule A W) := by
+  rintro ⟨_,_⟩
+  congr!
+
+instance : SetLike (Subrepresentation ρ) W where
+  coe ρ' := ρ'.toSubmodule
+  coe_injective' := SetLike.coe_injective.comp toSubmodule_injective
+
+def toRepresentation (ρ' : Subrepresentation ρ): Representation A G ρ'.toSubmodule where
+  toFun g := (ρ g).restrict (ρ'.apply_mem_toSubmodule g)
+  map_one' := by ext; simp
+  map_mul' x y := by ext; simp
+
+instance : Max (Subrepresentation ρ) where
+  max ρ₁ ρ₂ := .mk (ρ₁.toSubmodule ⊔ ρ₂.toSubmodule) <| by
+      simp [Submodule.forall_mem_sup]
```

Comment:
Can you remove the nonterminal `simp`?

Reply:
Are we happy with a non terminal "simp only"? I am guessing that the reason for asking for not having non terminal simps is to make proofs more robust to changes in the simp algorithm, but the "simp only"  case is some middle point which might be fine, depending on the community's preference

Reply:
Yes definitely! The rules are that after `simp` there are very few things which are allowed to happen (e.g. `ring`) but most of the time `simp` is supposed to be a closing tactic. Just replace `simp` with `simp?` and then the corresponding `simp only` (or see if you can remove the `simp` completely if you like).